### PR TITLE
YamlPmt improvements, safe type conversions, and lenient numeric parser

### PR DESCRIPTION
**Pull request by RalphSteinhagen**
**Summary of Changes:**
1. **YamlPmt improvements (first commit)**  
   - Replaced overly broad use of `auto` with explicit types for clarity and safer conversions.  
   - Fixed lifetime issues with `string_view::data()` by constructing proper `std::string` copies where needed.  
   - Ensured correct usage of `std::iscntrl`, `std::isspace`, etc. by casting to `unsigned char`, as required by the C++ standard to avoid undefined behaviour.  
   - Added a new helper function `formatAsLines(...)` to annotate lines and columns for YAML syntax errors, simplifying debugging.

   ```cpp
   // example snippet showing the new usage of formatAsLines:
   auto errorMsg = pmtv::yaml::formatAsLines(yamlInput, parseError);
   std::cerr << "Parse error:\n" << errorMsg << std::endl;
   ```
    **Example output:**

    ```text
    Unexpected exception with message:
    Could not parse yaml: 
     0:│⏎
     1:│blocks:⏎
     2:│  - name: main_source⏎
     3:│    id: good::fixed_source⏎
     4:│    parameters:⏎
     5:│      event_count: !!uint32 100
     6:│      unknown_property: !!int32_t aa⏎
                    Invalid value for type^
        
     7:│  - name: multiplier⏎
     8:│    id: good::multiply⏎
     9:│  - name: counter⏎
    10:│    id: builtin_counter⏎
    11:│  - name: sink⏎
    12:│    id: good::cout_sink⏎
    13:│    parameters:⏎
    14:│      total_count: 100
    15:│      unknown_property: 42⏎
    16:│⏎
    17:│connections:⏎
    18:│  - [main_source, 0, multiplier, 0]⏎
    19:│  - [multiplier, 0, counter, 0]⏎
    20:│  - [counter, 0, sink, 0]⏎
    21:│⏎
     at /home/rstein/git/graph-prototype/core/test/qa_grc.cpp:199
    ```
   
2. **Safe type conversions and parse-to-minimal-numeric utilities (second commit)**  
   - Introduced `convert_safely<T>(variant)`, which provides range-checked conversions among integrals, floats, enums, and partial support for complex types. Returns a `std::expected<T, std::string>` on success/failure.  
   - Added `parseToMinimalNumeric<Variant>(string_view)`, which reads a string, detects integer vs. float, and picks the smallest type in your `Variant` that can hold the parsed value. Negative integers skip unsigned checks; extremely large integers parse as `uint64_t` if it’s in the variant.  
   - Includes unit-tests (Boost.UT) to ensure correct behaviour for typical scenarios.

   ```cpp
   // snippet: safely_convert
   std::variant<int, float> var = 123;
   auto res = pmtv::convert_safely<float>(var);
   if (res) {
       // success: we have float(123.0f)
   } else {
       // error details: res.error()
   }

   // snippet: parseToMinimalNumeric
   using MyVariant = std::variant<int8_t, int64_t, float, double>;
   auto parsed = pmtv::parseToMinimalNumeric<MyVariant>("255");
   if (parsed) {
       // if MyVariant includes uint8_t, will store 255 as uint8_t
   }
   ```

3. **Lenient conversions in Settings (third commit)**  
   - Updated `Settings.hpp` to use `convert_safely` for type-safe and lenient assignment from PMT values to block parameters.  
   - Falls back to an exception if the provided value cannot be safely converted.  
   - Inlines a small code snippet:

   ```cpp
   if (auto convertedValue = pmtv::convert_safely<Type, strictConversion>(value); convertedValue) {
       newParameters.insert_or_assign(key, convertedValue.value());
   } else {
       throw std::invalid_argument(
           fmt::format("key '{}': cannot convert '{}': {}",
                       key, value.index(), convertedValue.error()));
   }
   ```

**Rationale:**
- Improves human readability and enforces explicit types, reducing hidden conversions.
- Fixes potential undefined behaviour in `std::isspace`/`std::isprint` calls.
- Allows easier debugging of YAML parse errors with line/column formatting.
- Strengthens type-safety through compile-time and runtime checks in the new “safe conversion” infrastructure.

### CODE REVIEW

File: core/include/gnuradio-4.0/PmtTypeHelpers.hpp [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+
+template<typename T>
+requires(std::is_floating_point_v<T>)
+static std::expected<T, std::string> parseStringToFloat(std::string_view trimmed) {
+    using namespace std::string_literals;
+#if defined(__clang__)
+    // Fallback to std::strtof / strtod for Clang versions prior to 20
+    if constexpr (std::is_same_v<T, float>) {
+        char* endPtr = nullptr;
+        float valF   = std::strtof(trimmed.data(), &endPtr);
```
Comment (drslebedev) [inline core/include/gnuradio-4.0/PmtTypeHelpers.hpp:89]: Why don't you use `std::std::stof`?

File: core/include/gnuradio-4.0/PmtTypeHelpers.hpp [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+            return valF;
+        }
+        if (std::isinf(valF)) {
+            return std::unexpected("float parse out-of-range"s);
+        }
+        return std::unexpected("invalid float parse"s);
+    } else {
+        // double
+        char*  endPtr = nullptr;
+        double valD   = std::strtod(trimmed.data(), &endPtr);
```
Comment (drslebedev) [inline core/include/gnuradio-4.0/PmtTypeHelpers.hpp:100]: same as above, why not `std::stod`?

File: core/include/gnuradio-4.0/PmtTypeHelpers.hpp [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+[[nodiscard]] constexpr R parseToMinimalNumeric(std::string_view numericString) {
+    const std::string_view str = detail::trimAndCutComment(numericString);
+    if (str.empty()) {
+        return std::unexpected(fmt::format("empty or comment-only string: '{}'", numericString));
+    }
+
+    const bool mightBeFloat = (str.find('.') != std::string_view::npos) || (str.find('e') != std::string_view::npos) || (str.find('E') != std::string_view::npos);
+
+    if (!mightBeFloat) { // attempt to parse as signed or unsigned integral
+        const bool isNegative = (!str.empty() && str.front() == '-');
```
Comment (drslebedev) [inline core/include/gnuradio-4.0/PmtTypeHelpers.hpp:362]: Just as an idea: should we also allow `+` sign, for example `+1234`?

**Review by drslebedev — APPROVED** [review]
Thanks for the PR! Great work on improving type-safety and overall code quality:
- I like  how you replaced broad `auto` usage with explicit types.
- Casting to `unsigned char` in `<cctype>` calls to avoid UB.
- The new safety conversion mechanism helps to keep yaml file simple!
- YAML error formatting with formatAsLines is a huge improvement for debugging.
    
The unit tests are well-done and thorough, covering integral boundary cases and complex conversions.

I also did minor comments in the code. But overall this looks ready to merge.

File: core/include/gnuradio-4.0/PmtTypeHelpers.hpp (hunk 1 above)
Comment (RalphSteinhagen) [reply core/include/gnuradio-4.0/PmtTypeHelpers.hpp:89]: works only with `std::string` and not `std::string_view' and also internally falls back to [std::strtof](https://en.cppreference.com/w/cpp/string/basic_string/stof)

File: core/include/gnuradio-4.0/PmtTypeHelpers.hpp (hunk 2 above)
Comment (RalphSteinhagen) [reply core/include/gnuradio-4.0/PmtTypeHelpers.hpp:100]: works only with `std::string` and not `std::string_view' and also internally falls back to [std::strtof](https://en.cppreference.com/w/cpp/string/basic_string/stof)

File: core/include/gnuradio-4.0/Settings.hpp [hunk 4]
```diff
@@ -1029,15 +1034,21 @@ class CtxSettings : public SettingsBase {
...
+                            return;
                         }
-                        if (fieldName == key && !std::holds_alternative<Type>(value)) {
-                            throw std::invalid_argument([&key, &value] { // lazy evaluation
+                        constexpr bool strictChecks = false;
+
+                        if (auto convertedValue = pmtv::convert_safely<Type, strictChecks>(value); convertedValue) [[likely]] {
+                            _stagedParameters.insert_or_assign(key, convertedValue.value());
+                            isSet = true;
+                            isSet = true;
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply core/include/gnuradio-4.0/Settings.hpp:1045]: Will remove this in the next PR. Not important here...

File: core/include/gnuradio-4.0/PmtTypeHelpers.hpp [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+template<std::unsigned_integral T>
+[[nodiscard]] constexpr bool inRange(std::int64_t val) {
+    return val < 0 ? false : static_cast<unsigned long long>(val) <= static_cast<std::uint64_t>(std::numeric_limits<T>::max());
+}
+
+constexpr std::string_view trimAndCutComment(std::string_view input) {
+    constexpr auto is_space = [](unsigned char ch) { return std::isspace(ch); };
+
+    // remove leading whitespace
+    input.remove_prefix(static_cast<std::size_t>(std::ranges::find_if_not(input, is_space) - input.begin()));
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply core/include/gnuradio-4.0/PmtTypeHelpers.hpp:68]: That's the reason why the `std::string` case isn't trimmed. All other cases must be trimmed for the subsequent functions to work.
